# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [1]:
#load python packages
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline
sns.set_style('darkgrid')

In [2]:
# show all columns
pd.set_option('display.max_columns', None)

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [3]:
path='C:\\Users\\chi\\Desktop\\spring_board\\Springboard\\Guided_capstone\\Step 2\\data'
os.chdir(path)

In [4]:
df = pd.read_csv('kmean_dfno.csv')

In [5]:
df.head()

Name       state  summit_elev  vertical_drop  \
0               Hilltop Ski Area      Alaska         2090            294   
1            Sunrise Park Resort     Arizona        11100           1800   
2  Yosemite Ski & Snowboard Area  California         7800            600   
3               Donner Ski Ranch  California         8012            750   
4                  June Mountain  California        10090           2590   

   trams  fastEight  fastSixes  fastQuads  quad  triple  double  surface  \
0      0        0.0          0          0     0       1       0        2   
1      0        0.0          0          1     2       3       1        0   
2      0        0.0          0          0     0       1       3        1   
3      0        0.0          0          0     0       1       5        2   
4      0        0.0          0          2     0       0       4        1   

   total_chairs  Runs  TerrainParks  LongestRun_mi  SkiableTerrain_ac  \
0             3  13.0           1.0            1.0               30.0   
1             7  65.0           2.0            1.2              800.0   
2             5  10.0           2.0            0.4               88.0   
3             8  52.0           2.0            1.5              505.0   
4             7  41.0           2.0            2.0             1500.0   

   Snow Making_ac  daysOpenLastYear  yearsOpen  averageSnowfall  AdultWeekday  \
0            30.0             150.0       36.0             69.0     30.000000   
1            80.0             115.0       49.0            250.0     74.000000   
2           100.0             110.0       84.0            300.0     47.000000   
3            60.0             163.0       82.0            400.0     75.000000   
4           100.0             114.0       58.0            250.0     57.916957   

   AdultWeekend  projectedDaysOpen  NightSkiing_ac  cluster  
0      34.00000              152.0            30.0        0  
1      78.00000              104.0            80.0        1  
2      47.00000              107.0             0.0        1  
3      75.00000              170.0             0.0        1  
4      64.16681              128.0             0.0        1

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [6]:
# drop first for baseline
dummies = pd.get_dummies(df['state'], drop_first= True)

In [7]:
dummies.head()

Arizona  California  Colorado  Connecticut  Idaho  Illinois  Indiana  Iowa  \
0        0           0         0            0      0         0        0     0   
1        1           0         0            0      0         0        0     0   
2        0           1         0            0      0         0        0     0   
3        0           1         0            0      0         0        0     0   
4        0           1         0            0      0         0        0     0   

   Maine  Maryland  Massachusetts  Michigan  Minnesota  Missouri  Montana  \
0      0         0              0         0          0         0        0   
1      0         0              0         0          0         0        0   
2      0         0              0         0          0         0        0   
3      0         0              0         0          0         0        0   
4      0         0              0         0          0         0        0   

   Nevada  New Hampshire  New Jersey  New Mexico  New York  North Carolina  \
0       0              0           0           0         0               0   
1       0              0           0           0         0               0   
2       0              0           0           0         0               0   
3       0              0           0           0         0               0   
4       0              0           0           0         0               0   

   Ohio  Oregon  Pennsylvania  Rhode Island  South Dakota  Tennessee  Utah  \
0     0       0             0             0             0          0     0   
1     0       0             0             0             0          0     0   
2     0       0             0             0             0          0     0   
3     0       0             0             0             0          0     0   
4     0       0             0             0             0          0     0   

   Vermont  Virginia  Washington  West Virginia  Wisconsin  Wyoming  
0        0         0           0              0          0        0  
1        0         0           0              0          0        0  
2        0         0           0              0          0        0  
3        0         0           0              0          0        0  
4        0         0           0              0          0        0

In [8]:
df = pd.concat([df, dummies], axis=1)
df.drop(columns =['state'], inplace = True)

In [9]:
df.head()

Name  summit_elev  vertical_drop  trams  \
0               Hilltop Ski Area         2090            294      0   
1            Sunrise Park Resort        11100           1800      0   
2  Yosemite Ski & Snowboard Area         7800            600      0   
3               Donner Ski Ranch         8012            750      0   
4                  June Mountain        10090           2590      0   

   fastEight  fastSixes  fastQuads  quad  triple  double  surface  \
0        0.0          0          0     0       1       0        2   
1        0.0          0          1     2       3       1        0   
2        0.0          0          0     0       1       3        1   
3        0.0          0          0     0       1       5        2   
4        0.0          0          2     0       0       4        1   

   total_chairs  Runs  TerrainParks  LongestRun_mi  SkiableTerrain_ac  \
0             3  13.0           1.0            1.0               30.0   
1             7  65.0           2.0            1.2              800.0   
2             5  10.0           2.0            0.4               88.0   
3             8  52.0           2.0            1.5              505.0   
4             7  41.0           2.0            2.0             1500.0   

   Snow Making_ac  daysOpenLastYear  yearsOpen  averageSnowfall  AdultWeekday  \
0            30.0             150.0       36.0             69.0     30.000000   
1            80.0             115.0       49.0            250.0     74.000000   
2           100.0             110.0       84.0            300.0     47.000000   
3            60.0             163.0       82.0            400.0     75.000000   
4           100.0             114.0       58.0            250.0     57.916957   

   AdultWeekend  projectedDaysOpen  NightSkiing_ac  cluster  Arizona  \
0      34.00000              152.0            30.0        0        0   
1      78.00000              104.0            80.0        1        1   
2      47.00000              107.0             0.0        1        0   
3      75.00000              170.0             0.0        1        0   
4      64.16681              128.0             0.0        1        0   

   California  Colorado  Connecticut  Idaho  Illinois  Indiana  Iowa  Maine  \
0           0         0            0      0         0        0     0      0   
1           0         0            0      0         0        0     0      0   
2           1         0            0      0         0        0     0      0   
3           1         0            0      0         0        0     0      0   
4           1         0            0      0         0        0     0      0   

   Maryland  Massachusetts  Michigan  Minnesota  Missouri  Montana  Nevada  \
0         0              0         0          0         0        0       0   
1         0              0         0          0         0        0       0   
2         0              0         0          0         0        0       0   
3         0              0         0          0         0        0       0   
4         0              0         0          0         0        0       0   

   New Hampshire  New Jersey  New Mexico  New York  North Carolina  Ohio  \
0              0           0           0         0               0     0   
1              0           0           0         0               0     0   
2              0           0           0         0               0     0   
3              0           0           0         0               0     0   
4              0           0           0         0               0     0   

   Oregon  Pennsylvania  Rhode Island  South Dakota  Tennessee  Utah  Vermont  \
0       0             0             0             0          0     0        0   
1       0             0             0             0          0     0        0   
2       0             0             0             0          0     0        0   
3       0             0             0             0          0     0        0   
4       0            

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [10]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the 
# result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign 
# it the AdultWeekend column of the df 
y = df.AdultWeekend 

# Here we use the StandardScaler() method of the preprocessing package,
# and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, 
# and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [11]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our 
# response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [12]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [13]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = lm.predict(X_test)

## Review Model Outcomes — Iterate over additional models as needed

In [14]:
# You might want to use the 
# explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics.
from sklearn.metrics import mean_absolute_error, explained_variance_score

# You can plug y_test and y_pred into the functions to evaluate the model
mae = mean_absolute_error(y_test, y_pred)
evs = explained_variance_score(y_test, y_pred)

In [15]:
print('Mean absolute error: ', str(mae))
print('Explained variance score: ', str(evs))

Mean absolute error:  6.641364743133798
Explained variance score:  0.5708410429134665


**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [16]:
print('Intercept: ', lm.intercept_)

Intercept:  56.87489571369453


**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [17]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
linear_coef= pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])
linear_coef.sort_values('Coefficient', ascending = False)

Coefficient
fastSixes          5.411886e+13
total_chairs       4.267683e+13
fastEight          3.450554e+13
trams              2.826004e+13
surface            2.432794e+13
double             2.295759e+13
triple             2.008801e+13
quad               1.214881e+13
fastQuads          9.661012e+12
AdultWeekday       1.057422e+01
New Hampshire      4.009766e+00
Vermont            3.447266e+00
Maine              3.292542e+00
New York           2.886963e+00
Massachusetts      2.738281e+00
North Carolina     2.699707e+00
West Virginia      2.544922e+00
Virginia           2.203125e+00
Pennsylvania       2.140625e+00
vertical_drop      2.129622e+00
Tennessee          1.850586e+00
cluster            1.592773e+00
yearsOpen          1.500000e+00
California         1.377930e+00
Wisconsin          1.348633e+00
New Jersey         1.208984e+00
Colorado           1.160156e+00
NightSkiing_ac     1.154297e+00
Maryland           1.085938e+00
daysOpenLastYear   1.072510e+00
Runs               1.058594e+00
New Mexico         1.058594e+00
Rhode Island       9.877930e-01
Michigan           9.687500e-01
Connecticut        9.349365e-01
Utah               9.101562e-01
SkiableTerrain_ac  8.691406e-01
Minnesota          8.378906e-01
projectedDaysOpen  7.890625e-01
Iowa               7.460938e-01
Montana            7.060547e-01
summit_elev        6.717336e-01
averageSnowfall    6.357422e-01
Illinois           6.328125e-01
Nevada             5.849609e-01
Idaho              5.800781e-01
Washington         4.882812e-01
TerrainParks       4.609375e-01
LongestRun_mi      3.027344e-01
Arizona            2.890625e-01
Missouri           2.880859e-01
Ohio               2.490234e-01
South Dakota       1.513672e-01
Oregon             1.083984e-01
Snow Making_ac     1.054688e-01
Wyoming            2.832031e-02
Indiana            8.789062e-03

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [18]:
df = pd.read_csv('kmean_dfno.csv')

In [19]:
X = df.drop(columns = ['Name', 'state', 'AdultWeekend'])
y = df['AdultWeekend']

In [20]:
scaler = preprocessing.StandardScaler().fit(X)
X_scaled= scaler.transform(X)

In [21]:
y = y.ravel()

In [22]:
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

In [23]:
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm2 = linear_model.LinearRegression()
model2 = lm2.fit(X_train,y_train)

In [24]:
from sklearn.metrics import mean_absolute_error, explained_variance_score
y_pred2= lm2.predict(X_test)

mae2 = mean_absolute_error(y_test, y_pred2)
evs2 = explained_variance_score(y_test, y_pred2)

In [25]:
print('Mean absolute error with state: ', mae)
print('Explained variance score with state: ', evs)
print('Mean absolute error without state: ', mae2)
print('Explained variance score without state: ', evs2)

Mean absolute error with state:  6.641364743133798
Explained variance score with state:  0.5708410429134665
Mean absolute error without state:  7.43189806435493
Explained variance score without state:  0.5161723660848809


In [26]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
linear_coef2= pd.DataFrame(abs(lm2.coef_), X.columns, columns=['Coefficient'])
linear_coef2.sort_values('Coefficient', ascending = False)

Coefficient
AdultWeekday       1.201812e+01
cluster            3.359305e+00
summit_elev        3.280186e+00
daysOpenLastYear   2.298188e+00
Runs               2.061892e+00
averageSnowfall    1.780114e+00
projectedDaysOpen  1.355180e+00
yearsOpen          9.606843e-01
double             9.080733e-01
quad               7.820465e-01
fastQuads          7.614638e-01
surface            6.252687e-01
NightSkiing_ac     4.131046e-01
Snow Making_ac     3.367506e-01
LongestRun_mi      3.040101e-01
total_chairs       1.586955e-01
SkiableTerrain_ac  1.270320e-01
vertical_drop      1.124038e-01
TerrainParks       1.037472e-01
triple             5.015311e-02
trams              4.884981e-15
fastEight          3.996803e-15
fastSixes          1.332268e-15

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number three spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

In [27]:
# base_elev was dropped at step3 due to high correlation

#### Model 3

In [28]:
df = pd.read_csv('kmean_dfno.csv')

In [29]:
X = df.drop(columns = ['Name', 'state', 'AdultWeekend', 'summit_elev'])
y = df['AdultWeekend']

In [30]:
scaler = preprocessing.StandardScaler().fit(X)
X_scaled= scaler.transform(X)

In [31]:
y = y.ravel()

In [32]:
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

In [33]:
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm3 = linear_model.LinearRegression()
model3 = lm3.fit(X_train,y_train)

In [34]:
from sklearn.metrics import mean_absolute_error, explained_variance_score
y_pred3= lm3.predict(X_test)

mae3 = mean_absolute_error(y_test, y_pred3)
evs3 = explained_variance_score(y_test, y_pred3)

In [35]:
print('Mean absolute error with state: ', mae)
print('Explained variance score with state: ', evs)
print('Mean absolute error without state: ', mae2)
print('Explained variance score without state: ', evs2)
print('Mean absolute error without state, summit_elev: ', mae3)
print('Explained variance score without state, summit_elev: ', evs3)

Mean absolute error with state:  6.641364743133798
Explained variance score with state:  0.5708410429134665
Mean absolute error without state:  7.43189806435493
Explained variance score without state:  0.5161723660848809
Mean absolute error without state, summit_elev, base_elev:  7.512617500970841
Explained variance score without state, summit_elev, base_elev:  0.5173943280267826


In [36]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
linear_coef3= pd.DataFrame(abs(lm3.coef_), X.columns, columns=['Coefficient'])
linear_coef3.sort_values('Coefficient', ascending = False)

Coefficient
AdultWeekday       1.143468e+01
cluster            2.982489e+00
averageSnowfall    2.887635e+00
daysOpenLastYear   1.968061e+00
Runs               1.909154e+00
double             1.001504e+00
projectedDaysOpen  9.587338e-01
yearsOpen          9.572024e-01
SkiableTerrain_ac  8.923049e-01
NightSkiing_ac     8.178935e-01
LongestRun_mi      5.297231e-01
fastQuads          4.768828e-01
quad               4.460415e-01
total_chairs       3.622838e-01
surface            3.225076e-01
Snow Making_ac     1.228941e-01
TerrainParks       6.024726e-02
vertical_drop      3.242389e-02
triple             2.472863e-02
trams              1.776357e-15
fastSixes          1.554312e-15
fastEight          1.332268e-15

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | 0.57 | 6.64 |-|
| Model 2. | 0.516 | 7.43 |'state'|
| Model 3. | 0.517 | 7.51 |'state','summit_elev'|

In [37]:
print('Mean absolute error with state: ', mae)
print('Explained variance score with state: ', evs)
print('Mean absolute error without state: ', mae2)
print('Explained variance score without state: ', evs2)
print('Mean absolute error without state, summit_elev: ', mae3)
print('Explained variance score without state, summit_elev: ', evs3)

Mean absolute error with state:  6.641364743133798
Explained variance score with state:  0.5708410429134665
Mean absolute error without state:  7.43189806435493
Explained variance score without state:  0.5161723660848809
Mean absolute error without state, summit_elev:  7.512617500970841
Explained variance score without state, summit_elev:  0.5173943280267826


#### Model Selection:
Model 1 is the best model for proving insights to Big Mountain management lift ticket prices. With the 'state' feature, the explained variance is higher than the other two models, which states that the model is telling more than the other two. Also, Mean absolute error is lower with 'state' feature.<br>
Even though controlling state variables is impossible, The 'state' feature cannot be ignored. The 'state' feature helps to segmentize and analyze which state is higher or lower-priced, and if one state has a too low priced lift, then other lifts, the management can decide to increase the price or consider a countermeasure.  